platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/LongListDisk.java

```diff
@@ -40,29 +90,103 @@ public class LongListDisk extends LongList {
-        this.file = file;
+        this(file, DEFAULT_RESERVED_BUFFER_LENGTH);
+    }
+
+    LongListDisk(final Path file, final long reservedBufferLength) throws IOException {
+        super(file, reservedBufferLength);
+        freeChunks = new ConcurrentLinkedDeque<>();
+    }
+
+    @Override
+    protected void readBodyFromFileChannelOnInit(String sourceFileName, FileChannel fileChannel) throws IOException {
+        Path tempFile = createTempFile(sourceFileName);
+        // create temporary file for writing
+        @SuppressWarnings("resource")
+        final RandomAccessFile rf = new RandomAccessFile(tempFile.toFile(), "rw");
+        // ensure that the amount of disk space is enough
+        // two additional chunks are required to accommodate "compressed" first and last chunks in the original file
+        rf.setLength(fileChannel.size() + (long) 2 * memoryChunkSize);
+        currentFileChannel = rf.getChannel();
+
+        final int totalNumberOfChunks = calculateNumberOfChunks(size());
+        final int firstChunkWithDataIndex = (int) (minValidIndex.get() / numLongsPerChunk);
+        final int minValidIndexInChunk = (int) (minValidIndex.get() % numLongsPerChunk);
+
+        // copy the first chunk
+        final ByteBuffer transferBuffer = initOrGetTransferBuffer();
+        transferBuffer.position(minValidIndexInChunk * Long.BYTES);
+        MerkleDbFileUtils.completelyRead(fileChannel, transferBuffer);
+        transferBuffer.flip();
+        // writing the full chunk, all values before minValidIndexInChunk are zeroes
+        MerkleDbFileUtils.completelyWrite(currentFileChannel, transferBuffer, 0);
+        chunkList.set(firstChunkWithDataIndex, 0L);
+
+        // copy everything except for the first chunk and the last chunk
+        MerkleDbFileUtils.completelyTransferFrom(
+                currentFileChannel, fileChannel, memoryChunkSize, (long) (totalNumberOfChunks - 2) * memoryChunkSize);
+
+        // copy the last chunk
+        transferBuffer.clear();
+        MerkleDbFileUtils.completelyRead(fileChannel, transferBuffer);
+        transferBuffer.flip();
+        MerkleDbFileUtils.completelyWrite(
+                currentFileChannel, transferBuffer, (long) (totalNumberOfChunks - 1) * memoryChunkSize);
+
+        for (int i = firstChunkWithDataIndex + 1; i < totalNumberOfChunks; i++) {
+            chunkList.set(i, (long) (i - firstChunkWithDataIndex) * memoryChunkSize);
+        }
+    }
+
+    private ByteBuffer initOrGetTransferBuffer() {
+        final ByteBuffer buffer;
+        if (TRANSFER_BUFFER_THREAD_LOCAL.get() == null) {
+            buffer = ByteBuffer.allocate(memoryChunkSize).order(ByteOrder.nativeOrder());
+            TRANSFER_BUFFER_THREAD_LOCAL.set(buffer);
+        } else {
+            buffer = TRANSFER_BUFFER_THREAD_LOCAL.get();
+            // clean up the buffer
+            buffer.clear();
+            Arrays.fill(buffer.array(), (byte) 0);
```

Comment:
This is a bit costly.  Do we really need to do this?  It's often cheaper to reallocate than write '0' to every location...

Reply:
Yes, you're right. In fact, it's needed for one specific case when we read incomplete first chunk from a snapshot. I used it there and removed it from here.

Reply:
I extracted zeroing out logic to `LongListDisk#fillBufferWithZeroes` method.